# Imports

In [4]:
import os, sys, glob, gc

from datetime import datetime, timedelta
import numpy as np

from read_matlab_data import DataFile
from organize_data import load_object


# Data Analysis

## File List

Loading all 2019 files that contain link information at 14.3644 MHz.

In [5]:
file_list = load_object("data\\2019\\2019_14p3644.dat")

## One File

In [8]:
data_file = file_list[0]
vars(data_file)

{'file_path': 'D:\\OneDrive - University of Calgary\\All Projects\\nrcan_data\\matlab_files\\2019\\ddAle_Ott_20190720_14364400.mat',
 'data_dic': {'antindf': array([[175, 175, 175, 175,   0, 175,   0, 175]], dtype=uint8),
  'azimuth': array([[191.50514604, 216.43512868, 211.61009272, 207.16897339,
           72.65682887, 184.41860091,  62.86301737,  53.66626571,
          202.27743117,  64.04606078,  60.62845587,  59.342674  ,
           60.04496692, 191.03062618,  70.53299976,  63.05381028,
           64.16061781,  64.22563088,  62.23702668, 213.42734996,
           63.72394069,  60.0725212 ,  91.65063039,  65.96469867,
           68.83500913,  62.34978763,  61.58136138,  63.18626882,
           62.75775923,  62.89187276,  69.61187221, 208.31466797,
           62.74712578,  71.31004583, 220.06533919,  59.24224167,
           60.56496753, 217.72333931,  65.07488526, 213.44083847,
           61.25749653,  62.57139038,  65.9210122 ,  61.26242995,
          208.92894678,  67.76700624, 214

## Time Data

In [9]:
date = data_file.date_time
print(date)

time_of_day = data_file.data_dic['timeday']
print(time_of_day)
print(time_of_day.shape)


2019-07-20 00:00:00
[[737626.69190972 737626.69885417 737626.70059028 737626.70232639
  737626.7040625  737626.70579861 737626.70753472 737626.70927083
  737626.71100694 737626.71274306 737626.71447917 737626.71621528
  737626.71795139 737626.7196875  737626.72142361 737626.72315972
  737626.72489583 737626.72663194 737626.72836806 737626.73010417
  737626.73184028 737626.73357639 737626.7353125  737626.73704861
  737626.73878472 737626.74052083 737626.74225694 737626.74399306
  737626.74572917 737626.74746528 737626.74920139 737626.7509375
  737626.75267361 737626.75440972 737626.75614583 737626.75788194
  737626.75961806 737626.76135417 737626.76309028 737626.76482639
  737626.7665625  737626.76829861 737626.77003472 737626.77177083
  737626.77350694 737626.77524306 737626.77697917 737626.77871528
  737626.78045139 737626.7821875  737626.78392361 737626.78565972
  737626.78739583 737626.78913194 737626.79086806 737626.79260417
  737626.79434028 737626.79607639 737626.7978125  737626.

In [10]:
time_of_day = time_of_day.ravel()
time_of_day

array([737626.69190972, 737626.69885417, 737626.70059028, 737626.70232639,
       737626.7040625 , 737626.70579861, 737626.70753472, 737626.70927083,
       737626.71100694, 737626.71274306, 737626.71447917, 737626.71621528,
       737626.71795139, 737626.7196875 , 737626.72142361, 737626.72315972,
       737626.72489583, 737626.72663194, 737626.72836806, 737626.73010417,
       737626.73184028, 737626.73357639, 737626.7353125 , 737626.73704861,
       737626.73878472, 737626.74052083, 737626.74225694, 737626.74399306,
       737626.74572917, 737626.74746528, 737626.74920139, 737626.7509375 ,
       737626.75267361, 737626.75440972, 737626.75614583, 737626.75788194,
       737626.75961806, 737626.76135417, 737626.76309028, 737626.76482639,
       737626.7665625 , 737626.76829861, 737626.77003472, 737626.77177083,
       737626.77350694, 737626.77524306, 737626.77697917, 737626.77871528,
       737626.78045139, 737626.7821875 , 737626.78392361, 737626.78565972,
       737626.78739583, 7

This suggests that everything before the decimal is just the day. The fraction might be the time of day, as the variable in the raw data is named `timeday`.


In [14]:
next_day_data =  file_list[1]
print(next_day_data.date_time)

n_time_data = next_day_data.data_dic['timeday'].ravel()
print(n_time_data)

2019-07-21 00:00:00
[737627.00267361 737627.00440972 737627.00614583 737627.00788194
 737627.00961806 737627.01135417 737627.01309028 737627.01482639
 737627.0165625  737627.01829861 737627.02003472 737627.02177083
 737627.02350694 737627.02524306 737627.02697917 737627.02871528
 737627.03045139 737627.0321875  737627.03392361 737627.03565972
 737627.03739583 737627.03913194 737627.04086806 737627.04260417
 737627.04434028 737627.04607639 737627.0478125  737627.04954861
 737627.05128472 737627.05302083 737627.05475694 737627.05649306
 737627.05822917 737627.05996528 737627.06170139 737627.0634375
 737627.06517361 737627.06690972 737627.06864583 737627.07038194
 737627.07211806 737627.07385417 737627.07559028 737627.07732639
 737627.0790625  737627.08079861 737627.08253472 737627.08427083
 737627.08600694 737627.08774306 737627.08947917 737627.09121528
 737627.09295139 737627.0946875  737627.09642361 737627.09815972
 737627.09989583 737627.10163194 737627.10336806 737627.10510417
 73762

This makes sense but what is the reference day...?

In [16]:
days_since = int(np.floor(n_time_data[0]))
days_since

737627

In [27]:
print(next_day_data.date_time)
print(datetime(1, 1, 1) + timedelta(days=days_since))

2019-07-21 00:00:00
2020-07-22 00:00:00


datetime.timedelta(days=201)

Okay. So their reference is year 0, day 0. The begining of time lol.

## Power Data

There's `snr`, `snr_dd`, `sigpow`, `peaktomean` and `hirespow` attributes in the data. All the data have 175 columns (can be though of as entries, and vary with each day and frequency- sometimes we have more/less data in a given day). However, all "hires" attributes, which includes `hirespow`, have dimensions (4 x 175)- 4 rows, 60 columns.

Luckily, a lot of the units here can be verified with Taylor's paper. From the paper, we have the following:

- SNR [dB]
- SIG POW [dBm] (referenced to milli-Watts)
- SNR_DD [dimensionless] (normalized probablility **d**ensity **d**istributions*) 


In [33]:
snr = data_file.data_dic['snr']

array([[4.18521542e+08, 3.38566836e+07, 9.03023567e+06, 4.26739645e+07,
        1.33771934e+07, 2.09698595e+07, 4.31008518e+07, 8.40452350e+07,
        1.41242543e+08, 7.21409284e+07, 8.79272319e+07, 1.58588692e+08,
        3.08420642e+08, 2.38515350e+08, 1.75694209e+08, 4.20824871e+08,
        2.24115012e+08, 3.55480411e+08, 1.88576897e+08, 1.98101385e+08,
        2.37917715e+08, 2.39370997e+08, 3.08843660e+08, 1.37406130e+08,
        1.69981038e+08, 2.15250382e+08, 3.98522167e+08, 2.80118935e+08,
        2.15695015e+08, 1.58493997e+08, 2.81434461e+08, 4.46101458e+08,
        2.89057305e+08, 5.64529147e+08, 3.34453333e+08, 2.11055102e+08,
        2.21003155e+08, 2.94474933e+08, 1.94254129e+08, 2.32180299e+08,
        1.82634368e+08, 4.06342702e+08, 3.15593482e+08, 5.15367385e+08,
        4.65795376e+08, 3.47821765e+08, 2.23947370e+08, 3.05641557e+08,
        1.88977258e+08, 2.73261713e+08, 2.83796451e+08, 3.03196282e+08,
        3.12559349e+08, 5.67863153e+08, 2.34781762e+08, 3.612923

### Signal Power

By inverting signal power, assuming the units were $dBm$, we can see that the mean value is about $ 4.6 \times 10^{5} \ kW $. In comparison, HAARP outputs ~ $3.6 \times 10^{3} ~ kW$ or ~ $ 95.5 ~ dBm $. It only makes sense that these are referenced to the noise floor, which is confirmed when looking at Taylor's paper: ***he used negative values***. 

In [43]:
sig_pow = 1e-3*np.power(data_file.data_dic['sigpow']/10, 10)
print(sig_pow[0, :3], "%.2e" % np.mean(sig_pow, axis=1))

[4.18521542e+08 3.38566836e+07 9.03023567e+06] 4.61e+08
[143.87129391]


### SNR_DD

These must be the density distributions that Taylor used in his paper. We can check by reproducing his 2014, 14.36 MHz plot. 

<img src="snr_dd_5mhz.png">

In [46]:
files_2014 = load_object("data\\2014\\2014_5p3820.dat")

In [48]:
file = files_2014[1]

file.data_dic['timeday']

array([[735601.011875  , 735601.02229167, 735601.03270833,
        735601.17854167, 735601.199375  , 735601.20979167,
        735601.22020833, 735601.230625  , 735601.24104167,
        735601.25145833, 735601.27229167, 735601.293125  ,
        735601.324375  , 735601.33479167, 735601.36604167,
        735601.37645833, 735601.386875  , 735601.45979167,
        735601.52229167, 735601.53270833, 735601.55354167,
        735601.886875  ]])

In [ ]:
from scipy import interpolate

time_arr = interpolate.interp1d()